Revit_Core_Engine/Convert/Environment/ToRevit/Duct.cs

```diff
@@ -0,0 +1,225 @@
+
+            SectionProfile sectionProfile = duct.SectionProperty?.SectionProfile;
+
+            DuctSectionProperty ductSectionProperty = duct.SectionProperty;
+            if(ductSectionProperty == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Duct creation requires a valid DuctSectionProperty.");
+            }
+
+            revitDuct = Duct.Create(document, mst.Id, ductType.Id, level.Id, start, end);
+            if (revitDuct == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("No Revit Duct has been created. Please check inputs prior to push attempt.");
+            }
+
+            // Copy parameters from BHoM object to Revit element
+            revitDuct.CopyParameters(duct, settings);
+
+            double orientationAngle = duct.OrientationAngle;
+            if (Math.Abs(orientationAngle) > settings.AngleTolerance)
+            {
+                ElementTransformUtils.RotateElement(document, revitDuct.Id, Line.CreateBound(start, end), orientationAngle);
+            }
+
+            double flowRate = duct.FlowRate;
+            revitDuct.SetParameter(BuiltInParameter.RBS_DUCT_FLOW_PARAM, flowRate);
+
+            double hydraulicDiameter = ductSectionProperty.HydraulicDiameter;
+            revitDuct.SetParameter(BuiltInParameter.RBS_HYDRAULIC_DIAMETER_PARAM, hydraulicDiameter);
+
+            // Rectangular Duct 
+            if (revitDuct.Shape() == ConnectorProfileType.Rectangular)
+            {
+                BoxProfile elementProfile = sectionProfile.ElementProfile as BoxProfile;
+                if (elementProfile == null)
+                    return null;
+
+                // Set Height
+                double profileHeight = elementProfile.Height;
+                revitDuct.SetParameter(BuiltInParameter.RBS_CURVE_HEIGHT_PARAM, profileHeight);
+
+                // Set Width
+                double profileWidth = elementProfile.Width;
+                revitDuct.SetParameter(BuiltInParameter.RBS_CURVE_WIDTH_PARAM, profileWidth);
+
+                // Set LiningProfile
+                BoxProfile liningProfile = sectionProfile.LiningProfile as BoxProfile;
+                if (liningProfile != null)
+                {
+                    double liningThickness = liningProfile.Thickness;
+
+                    // Get first available ductLiningType from document
+                    Autodesk.Revit.DB.Mechanical.DuctLiningType dlt = new FilteredElementCollector(document).OfClass(typeof(Autodesk.Revit.DB.Mechanical.DuctLiningType)).FirstOrDefault() as Autodesk.Revit.DB.Mechanical.DuctLiningType;
+                    if (dlt == null)
+                    {
+                        BH.Engine.Reflection.Compute.RecordError("Any duct lining type needs to be present in the Revit model in order to push ducts with lining.");
+                        return null;
```

Comment:
you either need to do this check before the creation of a duct or this should not break the procedure (now we could end  up with a duct created, but null returned)

Reply:
Changing this to record warning and not return null if that works. Since we need this check if there is a lining profile present.